dev/NavigationView/NavigationView.cpp

```diff
@@ -379,10 +427,354 @@ void NavigationView::OnApplyTemplate()
+    // when clicking overflow on topnav, transition is from bottom
+    // otherwise if prevItem is on left side of nextActualItem, transition is from left
+    //           if prevItem is on right side of nextActualItem, transition is from right
+    // click on Settings item is considered Default
+    NavigationRecommendedTransitionDirection recommendedDirection = NavigationRecommendedTransitionDirection::Default;
+    if (IsTopNavigationView() && itemSelectsOnInvoked)
+    {
+        if (isInOverflow)
+        {
+            recommendedDirection = NavigationRecommendedTransitionDirection::FromOverflow;
+        }
+        else if (prevItem && nvi)
+        {
+            recommendedDirection = GetRecommendedTransitionDirection(NavigationViewItemBaseOrSettingsContentFromData(prevItem), nvi);
+        }
+    }
+
+    RaiseItemInvoked(nextItem, false /*isSettings*/, nvi, recommendedDirection);
+}
+
+void NavigationView::OnNavigationViewItemInvoked(const winrt::NavigationViewItem& nvi)
+{
+    auto selectedItem = SelectedItem();
+    RaiseItemInvokedForNavigationViewItem(nvi);
+    auto selectedItemAfterInvokeCallback = SelectedItem();
+
+    // User changed selectionstate in the ItemInvoked callback
+    if (selectedItem != selectedItemAfterInvokeCallback)
+    {
+        return;
+    }
+
+    bool itemSelectsOnInvoked = nvi.SelectsOnInvoked();
+    if (m_selectionModel && itemSelectsOnInvoked)
+    {
+        winrt::IndexPath ip = GetIndexPathForContainer(nvi);
+        m_selectionModel.SelectAt(ip);
+    }
+
+    ClosePaneIfNeccessaryAfterItemIsClicked();
+}
+
+bool NavigationView::IsRootItemsRepeater(winrt::hstring name)
+{
+    return (name == c_topNavRepeater ||
+        name == c_leftRepeater ||
+        name == c_overflowRepeater);
+}
+
+winrt::FrameworkElement NavigationView::GetParentForFrameworkElement(winrt::FrameworkElement fe)
+{
+    auto parent = fe.Parent().try_as<winrt::FrameworkElement>();
+    if (!parent)
+    {
+        parent = (winrt::VisualTreeHelper::GetParent(fe)).try_as<winrt::FrameworkElement>();
+    }
+    return parent;
+}
+
+winrt::ItemsRepeater NavigationView::GetParentItemsRepeaterForContainer(winrt::NavigationViewItemBase nvib)
+{
+    auto child = nvib.try_as<winrt::FrameworkElement>();
+    auto parent = GetParentForFrameworkElement(child);
+
+    if (parent != nullptr)
+    {
+        if (auto parentIR = parent.try_as<winrt::ItemsRepeater>())
+        {
+            return parentIR;
+        }
+    }
+    return nullptr;
+}
+
+winrt::IndexPath NavigationView::GetIndexPathForContainer(winrt::NavigationViewItemBase nvib)
+{
+    auto path = std::vector<int>();
+
+    auto child = (nvib).try_as<winrt::FrameworkElement>();
+    auto parent = GetParentForFrameworkElement(child);
+    if (parent == nullptr)
+    {
+        return IndexPath::CreateFromIndices(path);
+    }
+
+    while (!(parent.try_as<winrt::ItemsRepeater>()) || !IsRootItemsRepeater((parent.try_as<winrt::ItemsRepeater>()).Name()))
```

Comment:
you can simplify this condition if you update IsRootItemsRepeater to take a UIElement instead. so this will be just
while(!IsRootItemsRepeater(parent))